### ISSUE TITLE
Why did we need mix?

### ISSUE THREAD

**Issue by rlewkowicz** [JeffreyWay/laravel-mix#369]
I don't know if this is the correct place for this question, but why this project? I don't mean that as a rude question. You're contributing to a very intelligent framework, and I'm honestly interested in why mix was decided to be needed. I ask this as someone who never really got past beginner mode about a year ago and I'm now coming back. First there was Laravel Elixer which "simplified" gulp. Now you have Laravel Mix which "simplifies" web pack. What was wrong with elixer? Even more whats wrong with just gulp? I want to use the built in facilities to Laravel. I get that I can hack my own into place, but then I mind as well just use lumen and go from there. It doesn't come across as friendly. This is great if I want to watch one file and compile that to one output file. Who does that? Most people will have at least some directory structure for their css/js and this does not support wild cards/globbing so far as I can tell. Now I'm going to try this (and adapt it to mix):
http://stackoverflow.com/questions/35098935/laravel-elixir-use-wildcard-without-combining/39673313#39673313
and I don't even need multiple output files. I just don't want to manually enter a path for each mix. I feel like that's gross. Then we get into why are you even doing this? What was so fundamentally broken that this change was required?

**Comment by nealstammers** [comment]
Laravel Mix takes a slightly different path to Elixir, hence a 'new' platform. Both products greatly simplify the process of building all your site assets.

Both tools can watch multiple files being changed, so you shouldn't need to specify every single one. But for example, if you're using SASS you should have one 'main' file in which you specify all other SASS files to include. Then, use Elixir or Mix to watch and compile just the 'main' SASS file. The great thing is that if ANY of the SASS files change, it will be re-compiled.

I'm not sure if this is a free video, but if you're a laracaster you could watch this for more info:
https://laracasts.com/series/whats-new-in-laravel-5-4/episodes/3

The forum would be a good place to add questions like this on too.
https://laracasts.com/discuss

Cheers, N.

**Comment by rlewkowicz** [comment]
I guess the question was whats the broader vision? There's always an objective. What problem did this aim to solve, or whats the ideological objective? 

As for the multiple files, I like this: https://sass-guidelin.es/#architecture

Lets say I have an entry point for my sass. app.scss. I and I have that in a root dir assets/sass. 

so I have 
app.scss
folder1/partial.scss
folder2/partial.scss
folder3/partial.scss
and so on...

So with gulp I could do:
gulp.watch('resources/assets/stylus/**/*.styl', ['stylus']);
and
gulp.src('resources/assets/stylus/style.styl')


and then my includes looked like:

@import "base/all.styl"
@import "components/all.styl"
@import "layout/all.styl"

Mix fails if I try this. It's like it doesn't support wildcards (globbing?). With everything in the same dir, with local level include, it works. I don't know if it's a limitation of sass (I abandoned stylus just so I could work within this frame work. Cause yea right trying to use non standard pre and post processors with mix), or mix. I could go dig through documentation and browse what few forum posts there are (this is pretty new and google mixes it up with elixer). Maybe poke at the code. But then how is this any easier? And then why did we make this design choice?

**Comment by adiachenko** [comment]
@rlewkowicz I haven't used Elixir much, but from my experience with creating build pipelines, the problems are as follows:

- Elixir still use Webpack under the hood (surprise) by utilizing [webpack-stream](https://github.com/shama/webpack-stream) interface that is no longer maintained and has a [plethora of issues on it's own](https://github.com/shama/webpack-stream/pull/126)
- Elixir had way too many dependencies. There was a lot of cases prior to version 5 or 6 (before browserify setup was simplified and then ditched entirely) where it didn't even install on some systems (Windows being the main suspect) due some dependency failing for some obscure reason
- Elixir was too hard to meaningfully customize. There was no single point of truth like the current `setup/webpack.config.js` that you can tweak to your needs
- Elixir was too hard to maintain. `laravel-mix` has much more elegant architecture that is easy for understand for everyone who had some experience with webpack while the previous gulp setup was a voodoo with dozens of plugins.

Anyway, I am glad that Laravel officially ditched that Gulp monstrosity and I can sleep soundly knowing I don't need to deal with it's issues or maintain my own webpack boilerplate. I'm already using laravel-mix and I have a much better experience with it

**Comment by JeffreyWay** [comment]
> Even more whats wrong with just gulp?

Well, for that matter, why do you even need Gulp? Just wire it up all yourself. 

>  There's always an objective. What problem did this aim to solve, or whats the ideological objective?

 Maybe build up a Gulpfile from scratch that does everything Mix does, and then let's compare your version versus the webpack.mix.js file that ships with Laravel.

> Cause yea right trying to use non standard pre and post processors with mix), or mix. 

I don't know what you mean by non-standard pre-processors. There's no need for you to pass a glob to fetch all Sass files within a directory. You can't ensure the order, and it's just not needed. Point to a single entry Sass file, and, there, you can import any partials that it requires.

**Comment by rlewkowicz** [comment]
That what I'm saying though. I can't even figure out how to wire up mix to do the same thing. 

There's this guy:
https://laracasts.com/discuss/channels/elixir/anyone-know-how-to-use-new-laravel-mix-with-stylus-and-post-processors
Who's like go edit some files. 

How would I use stylus and lost grid? (Which I picked up from laracasts) 

Which is all I found so far. 

I think I figured out how to do my sass includes with mix though
